src/v/cluster/partition.cc

```diff
@@ -1213,40 +1213,90 @@ partition::get_cloud_storage_manifest_view() {
 ss::future<result<model::offset, std::error_code>>
 partition::sync_kafka_start_offset_override(
   model::timeout_clock::duration timeout) {
+    model::offset offset_override;
+
     if (_log_eviction_stm && !is_read_replica_mode_enabled()) {
         auto offset_res
           = co_await _log_eviction_stm->sync_start_offset_override(timeout);
         if (offset_res.has_failure()) {
             co_return offset_res.as_failure();
         }
+
+        offset_override = offset_res.value();
+
```

Comment:
I tried taking a step back to understand how this looks / should look with these new cached variables. I think this is roughly what we have here, but there are some nuances. It seems like the cached values are basically caching the translation, but in a potentially imperfect way.

```
raft_start_override = co_await sync_log_eviction_stm()
if raft_start_override.has_value():
    if raft_start_override == raft_start_cache:
        # Easy case, we've translated before.
        return kafka_start_cache

    # We haven't translated before, but data is still in local log.
    if we can translate:
        translated_override = translate(raft_start_override)
        raft_start_cache = raft_start_override
        kafka_start_cache = translated_override

    # Can't translate because a race has likely occurred: start override has
    # fallen out of local log, and there is no cached value to fall back on.
    # Fall through.

# We may or may not have already synced above, but we sync again to ensure the
# value we get from archival is up-to-date.

kafka_start_override = co_await sync_archival_stm()

if raft_start_override.has_value():
    # Note that even if there is an override from the log_eviction_stm AND from
    # the archival_stm, they don't necessarily correspond to the same record
    # (e.g. consider a race where DeleteRecords comes in after the above sync).
    #
    # If this is the case though, subsequent calls will resync the log eviction
    # stm and not use this cached value.
    raft_start_cache = raft_start_override
    kafka_start_cache = kafka_start_override

return kafka_start_override
```

Reply:
Lmk if this is more or less what you're thinking here. If so, maybe we can incorporate the structure or comments into the implementation?

Reply:
I've hopefully simplified the caching logic and added clearer comments in the updated version as per our conversations on Slack. Let me know if they're still lacking though.